From 0a0b2fb5eab4000db45c9d5827b0fd3bd6190010 Mon Sep 17 00:00:00 2001
From: Sriram Dash <quic_sriramd@quicinc.com>
Date: Mon, 8 Jul 2024 11:42:24 +0530
Subject: [PATCH 05/17] phy: qcom: qmp-usb: Call qmp_usb_remove during driver
 remove

This driver users the component model and the clean up of resources
should happen during remove.

Ported from meta-qti-auto-kernel patch: 0010-PENDING-phy-qcom-qmp-usb-Call-qmp_usb_remove-during-.patch

Signed-off-by: Sriram Dash <quic_sriramd@quicinc.com>
Signed-off-by: Shazad Hussain <quic_shazhuss@quicinc.com>
---
 drivers/phy/qualcomm/phy-qcom-qmp-usb.c | 12 ++++++++++++
 1 file changed, 12 insertions(+)

diff --git a/drivers/phy/qualcomm/phy-qcom-qmp-usb.c b/drivers/phy/qualcomm/phy-qcom-qmp-usb.c
index b0790bcf0bc8..03b0c320d6cd 100644
--- a/drivers/phy/qualcomm/phy-qcom-qmp-usb.c
+++ b/drivers/phy/qualcomm/phy-qcom-qmp-usb.c
@@ -2467,6 +2467,17 @@ static int qmp_usb_probe(struct platform_device *pdev)
 	return ret;
 }
 
+static void qmp_usb_remove(struct platform_device *pdev)
+{
+	struct qmp_usb *qmp = platform_get_drvdata(pdev);
+	const struct qmp_phy_cfg *cfg = qmp->cfg;
+
+	clk_disable_unprepare(qmp->pipe_clk);
+	reset_control_bulk_assert(qmp->num_resets, qmp->resets);
+	clk_bulk_disable_unprepare(qmp->num_clks, qmp->clks);
+	regulator_bulk_disable(cfg->num_vregs, qmp->vregs);
+}
+
 static const struct of_device_id qmp_usb_of_match_table[] = {
 	{
 		.compatible = "qcom,glymur-qmp-usb3-uni-phy",
@@ -2532,6 +2543,7 @@ MODULE_DEVICE_TABLE(of, qmp_usb_of_match_table);
 
 static struct platform_driver qmp_usb_driver = {
 	.probe		= qmp_usb_probe,
+	.remove		= qmp_usb_remove,
 	.driver = {
 		.name	= "qcom-qmp-usb-phy",
 		.pm	= &qmp_usb_pm_ops,
-- 
2.34.1

